### PR TITLE
Done filtering using a map changes

### PR DESCRIPTION
**Pull request by shivani-0409**

### CODE REVIEW

File: api/handlers/find_user.go [hunk 1]
```diff
@@ -16,21 +15,28 @@ type findUser struct{
...
 	usResponse:=[]*models.User{}
 	for _,usr:=range us{
 		usResponse=append(usResponse,asUserResponse(usr))
 	}
 	res:=users.NewFindUsersOK().WithPayload(usResponse)
 	return res
-	// print the appication name
 }
-
+func filteredMap(Name *string) map[string]interface{}{
```
Comment (sewak2610) [inline api/handlers/find_user.go]: pass `params FindUserParams` here 
and access Name like params.Name inside the function.

File: api/handlers/find_user.go [hunk 2]
```diff
@@ -16,21 +15,28 @@ type findUser struct{
...
-	
-	fmt.Println("Application Name is",f.rt.GetApplicationName())
-	// n:="Shivani Sharma"
-	// us:= []*models.User{{Address:"ABC",ID:2,Name:&n}}
-	
-	us,_:= f.rt.GetManager().ListUsers(fup.Name,fup.Limit)
+
+	var limit int32
+	// var name string
+	var us []*domain.User
```
Comment (sewak2610) [inline api/handlers/find_user.go]: not required. use initialization concept like 
us, err := <func call >

File: api/handlers/find_user.go [hunk 3]
```diff
@@ -16,21 +15,28 @@ type findUser struct{
...
+	// if fup.Limit != nil && fup.Name != nil {
+		limit = *fup.Limit
+		// name = *fup.Name
+		us,_ = f.rt.GetManager().ListUsers(filteredMap(fup.Name), limit)
```
Comment (sewak2610) [inline api/handlers/find_user.go]: need to handle err as well 
it should be looks 
`us, err :== f.rt.GetManager().ListUsers(filteredMap(fup.Name), limit)`

File: api/handlers/find_user.go [hunk 4]
```diff
@@ -16,21 +15,28 @@ type findUser struct{
...
+	// }else{
+	// 	us,_ = f.rt.GetManager().ListUsers("", limit)
```
Comment (sewak2610) [inline api/handlers/find_user.go]: remove this commented code.

File: api/handlers/find_user.go (hunk 1 above)
Comment (sewak2610) [inline api/handlers/find_user.go]: function name `extractFilters` will be more suitable.

File: db/inmemory/client.go [hunk 5]
```diff
@@ -4,6 +4,7 @@ import(
 	"github.com/go-swagger/go-swagger/examples/ShivaniCustomServerExample1/domain"
 	"github.com/go-swagger/go-swagger/examples/ShivaniCustomServerExample1/db"
 	// "github.com/segmentio/ksuid"
```
Comment (sewak2610) [inline db/inmemory/client.go]: please remove commented code all over the palaces.
in case it has some significance then add a comment why??...

File: db/inmemory/client.go [hunk 6]
```diff
@@ -24,22 +25,19 @@ func (c *client) AddUser(user *domain.User) (string, error){
 	c.ds[user.ID]=user
 	return user.ID,nil
 }
-func (c *client) ListUsers(queryName *string, queryLimit *int32) ([]*domain.User,error){
+func (c *client) ListUsers(queryNameMap map[string]interface{}, queryLimit int32) ([]*domain.User,error){
```
Comment (sewak2610) [inline db/inmemory/client.go]: instead of `queryNameMap` use `filters`
and `queryLimit` will be `limit` only

File: db/inmemory/client.go [hunk 7]
```diff
@@ -24,22 +25,19 @@ func (c *client) AddUser(user *domain.User) (string, error){
...
 	var userInfo =[]*domain.User{}
-	if queryName != nil && queryLimit !=nil{
-		fmt.Println("The queryName and queryLimit are",*queryName,*queryLimit)
-	}
 	var lim int32 = 1
 	for _,value := range c.ds{
-		if queryName !=nil{
-			if *queryName == value.Name{
+		if queryNameMap["name"] !=nil{
+			if strings.ToUpper(fmt.Sprint(queryNameMap["name"])) == strings.ToUpper(value.Name){
```
Comment (sewak2610) [inline db/inmemory/client.go]: what is the purpose to convert to upper here?

File: db/mongo/client.go [hunk 8]
```diff
@@ -60,101 +61,71 @@ func (c *client) AddUser(user *domain.User) (string, error){
 	//fmt.Println("The ID is",id)
 	return user.ID,nil
 }
-func (c *client) ListUsers(queryName *string,queryLimit *int32) ([]*domain.User,error){
+func (c *client) ListUsers(queryNameMap map[string]interface{},queryLimit int32) ([]*domain.User,error){
```
Comment (sewak2610) [inline db/mongo/client.go]: same comment as in inmemory.

File: db/mongo/client.go [hunk 9]
```diff
@@ -60,101 +61,71 @@ func (c *client) AddUser(user *domain.User) (string, error){
...
-		var result bson.D
+	for cur.Next(context.Background()) {
 		var result1 *domain.User	
 		err := cur.Decode(&result1)
 		if err != nil { log.Fatal(err) }
 		resultArray=append(resultArray,result1)
-		fmt.Println("The result is",result)
 	}
 	if err := cur.Err(); err != nil {
 		log.Fatal(err)
```
Comment (sewak2610) [inline db/mongo/client.go:75]: here
return nil, err 
after the log

File: db/mongo/client.go [hunk 10]
```diff
@@ -60,101 +61,71 @@ func (c *client) AddUser(user *domain.User) (string, error){
...
-	// if err := cur.Err(); err != nil {
-	// 	log.Fatal(err)
-	// }
 	return &userInfo,nil
+}
+
+func applyFilter(filterMap map[string]interface{}) map[string]interface{} {
+    for k, v := range filterMap {
+        switch mval := v.(type) {
+        case []string:
```
Comment (sewak2610) [inline db/mongo/client.go:98]: what is the purpose of this case?

File: db/mongo/client.go [hunk 11]
```diff
@@ -60,101 +61,71 @@ func (c *client) AddUser(user *domain.User) (string, error){
...
+            filterMap[k] = bson.M{"$in": mval}
+        case string:
+            // support searching by name using case-insensitive matching
+            if k == "name" {
+                filterMap[k] = bson.M{"$regex": primitive.Regex{Pattern: "^" + regexp.QuoteMeta(mval) + "$", Options: "i"}}
```
Comment (sewak2610) [inline db/mongo/client.go:103]: what do you understand from this regex?

File: runtime.go [hunk 12]
```diff
@@ -7,7 +7,7 @@ import(
 func NewRunTime(an string) (*Runtime){
 	return &Runtime{
 		AppName:an,
-		svc: service.NewManager("mongo"),
+		svc: service.NewManager("inmemory"),
```
Comment (sewak2610) [inline runtime.go]: rather than hard-coding here. 
read from runtime as we do for `port` and default will `mongo`
But, best practices are read from the environment variable, for that you can use viper.
https://github.com/spf13/viper
FYI it will not be part of this pull request.

File: service/interface.go [hunk 13]
```diff
@@ -15,7 +15,7 @@ import(
 )
 type Manager interface{
 	CreateUser(*domain.User) (string,error)
```
Comment (sewak2610) [inline service/interface.go]: although, the parameter's names are optional here,
But if you add those, it will increase readability

File: service/manager.go [hunk 14]
```diff
@@ -16,63 +16,19 @@ type manager struct{
...
-// 		if(value == m.ds[i]){
-// 			delete(m.ds,i)
-// 			return nil
-// 		}
-// 	  }
-// 	  message:="User is not found"
-// 	  return &models.Error{Code:404,Message:&message}
-// }
-
 func (m *manager) DeleteUser(i string) error{
```
Comment (sewak2610) [inline service/manager.go]: instead of `i` you can use `id`

File: service/manager.go [hunk 15]
```diff
@@ -16,63 +16,19 @@ type manager struct{
...
-	// if _, ok := m.ds[i]; ok{
-	// 		  delete(m.ds,i)
-	// 		   return nil
-	// 	}
-	// 	 return &domain.Error{Code:404,Message:"User doesn't exist"}
 	_,err:=m.ds.ViewUser(i)
```
Comment (sewak2610) [inline service/manager.go]: this code can be inline like
```
if _,err:=m.ds.ViewUser(id); err != nil{
		return &domain.Error{Code:404,Message:"User doesn't exist"}
	}
return  m.ds.DeleteUser(id)	
```

File: service/manager.go [hunk 16]
```diff
@@ -85,10 +41,6 @@ func (m *manager) DeleteUser(i string) error{
 
   func (m *manager) ViewUser(i string) (*domain.User,error){
```
Comment (sewak2610) [inline service/manager.go]: rename argument to `id`

File: swagger/swagger.yml [hunk 17]
```diff
@@ -47,7 +47,7 @@ paths:
           in: query
           type: integer
           format: int32
```
Comment (sewak2610) [inline swagger/swagger.yml]: if you remove `format: int32`
then you don't need to type-cast in code.

**Review by sewak2610 — CHANGES_REQUESTED** [review]

File: api/handlers/find_user.go (hunk 1 above)
Comment (shivani-0409) [reply api/handlers/find_user.go]: Done ! Please check

File: api/handlers/find_user.go (hunk 2 above)
Comment (shivani-0409) [reply api/handlers/find_user.go]: Done ! Please check

File: api/handlers/find_user.go (hunk 3 above)
Comment (shivani-0409) [reply api/handlers/find_user.go]: Done ! Please check

File: api/handlers/find_user.go (hunk 4 above)
Comment (shivani-0409) [reply api/handlers/find_user.go]: Done ! Please check

File: api/handlers/find_user.go (hunk 1 above)
Comment (shivani-0409) [reply api/handlers/find_user.go]: Done ! Please check

File: db/inmemory/client.go (hunk 5 above)
Comment (shivani-0409) [reply db/inmemory/client.go]: Okay. Removed the commented lines

File: db/inmemory/client.go (hunk 6 above)
Comment (shivani-0409) [reply db/inmemory/client.go]: Done ! Please check

File: db/inmemory/client.go (hunk 7 above)
Comment (shivani-0409) [reply db/inmemory/client.go]: Sir if user tries to search a name in any case(upper, lower or mixed) then also he/she should get the filtered object for that name. That's why did like this. Please suggest if this is not supposed to be done or something better can be done.

File: db/mongo/client.go (hunk 8 above)
Comment (shivani-0409) [reply db/mongo/client.go]: Done ! Please check

File: db/mongo/client.go (hunk 9 above)
Comment (shivani-0409) [reply db/mongo/client.go:75]: Done ! Please check

File: db/mongo/client.go (hunk 10 above)
Comment (shivani-0409) [reply db/mongo/client.go:98]: Sir currently it is not in use, but for suppose we want to search for multiple names at a time, then it can be used to get those records.

File: db/mongo/client.go (hunk 11 above)
Comment (shivani-0409) [reply db/mongo/client.go:103]: Sir this is mainly used for searching pattern in a text/string. Suppose user wants to search a name, so with regexp or regex he/she can use any special chars in the text or spaces; upper/lower/mixed case; incomplete text. But using regex all the matching records will be filtered out.

File: runtime.go (hunk 12 above)
Comment (shivani-0409) [reply runtime.go]: Okay Sir, will try this.

File: service/interface.go (hunk 13 above)
Comment (shivani-0409) [reply service/interface.go]: Done ! Please check

File: service/manager.go (hunk 14 above)
Comment (shivani-0409) [reply service/manager.go]: Done ! Please check

File: service/manager.go (hunk 15 above)
Comment (shivani-0409) [reply service/manager.go]: Done ! Please check

File: swagger/swagger.yml (hunk 17 above)
Comment (shivani-0409) [reply swagger/swagger.yml]: Done ! Please check

File: api/handlers/delete_user.go [hunk 18]
```diff
@@ -13,24 +11,10 @@ type deleteUser struct{
...
-// 	if isError == nil{
-// 		res:=users.NewDeleteUserNoContent()
-// 		return res
-// 	}
-// 	res:=users.NewDeleteUserNotFound().WithPayload(isError)
-// 	return res
-// }
-
 func (d *deleteUser) Handle(del users.DeleteUserParams) middleware.Responder{
 	err :=d.rt.GetManager().DeleteUser(del.ID)
```
Comment (sewak2610) [inline api/handlers/delete_user.go]: this can be simplify with inline 
if err := <calll>; err!nil{
// code block 
}

File: api/handlers/find_user.go [hunk 19]
```diff
@@ -1,36 +1,34 @@
...
-	// v.ListUsers()
-	
-	fmt.Println("Application Name is",f.rt.GetApplicationName())
-	// n:="Shivani Sharma"
-	// us:= []*models.User{{Address:"ABC",ID:2,Name:&n}}
-	
-	us,_:= f.rt.GetManager().ListUsers(fup.Name,fup.Limit)
+	us,err := f.rt.GetManager().ListUsers(extractFilters(fup), *fup.Limit)
+	if err != nil{
+		log.Fatal(err)
```
Comment (sewak2610) [inline api/handlers/find_user.go]: rather log.Fatal here 
log the error 
and return default response by append err.Message in message block.

File: db/client.go [hunk 20]
```diff
@@ -2,14 +2,13 @@ package db
...
-	// "github.com/go-swagger/go-swagger/examples/ShivaniCustomServerExample1/db/inmemory"
 )
 
 type DataStore interface{
-	AddUser(*domain.User) (string, error)
-	ListUsers(*string,*int32) ([]*domain.User,error)
-	DeleteUser(string) error
-	ViewUser(string) (*domain.User,error)
+	AddUser(user *domain.User) (string, error)
+	ListUsers(mapNames map[string]interface{},limit int64) ([]*domain.User,error)
```
Comment (sewak2610) [inline db/client.go]: `mapNames` can be renamed to `filters`

File: db/mongo/client.go (hunk 10 above)
Comment (sewak2610) [reply db/mongo/client.go:98]: yes, if interface is type of []string then it will treat as IN in db query.

File: db/mongo/client.go (hunk 11 above)
Comment (sewak2610) [reply db/mongo/client.go:103]: yes, 
One more thing, mongo doesn't support `LIKE` constraint, so we can achieve it with regex.

**Review by sewak2610 — CHANGES_REQUESTED** [review]
commit message should not be change
it will remain same, I saw you updated 
revert commit message to `Done filtering using a map changes` or whatever

File: runtime.go [hunk 21]
```diff
@@ -34,3 +35,26 @@ func (rt *Runtime)GetManager() service.Manager{
 	return rt.svc
 }
 
+func EnvironmentSpecificConfig(v *viper.Viper, env string) *viper.Viper {
```
Comment (sewak2610) [inline runtime.go]: where are you using that??
it shouldn't be part of this pull request, 
revert back.

**Review by sewak2610 — COMMENTED** [review]
Minor change. 
I am assuming you tested on local, it is working.

**Comment by sewak2610** [comment]
LGTM !! you can merge this pull request.
